zio-http/src/main/scala/zio/http/netty/client/ConnectionPool.scala

```diff
@@ -148,6 +153,7 @@ object ConnectionPool {
 
   private class ZioConnectionPool(
     pool: ZKeyedPool[Throwable, PoolKey, JChannel],
+    activeChannels: Ref[Set[JChannel]],
```

Comment:
Why isn't this being handled by the pool? Is the pool not being eager enough in releasing resources?

Reply:
There is an external event (the channel's "closed" future) that we want to listen for, and invalidate the channel in the pool when it happens - for example the channel was closed by a network error, etc.

The memory leak was that a new listener was added to this event on every `get` from the pool - but we only want to do it once.

Maybe it would be better to add this into `createChannel` but then we need to pass a promise of the pool to it. I could not decide which one is better.

Reply:
I would say the `Promise` option (actually you only need to pass `Promise#await`, not the whole `Promise`). Because this one has significant overhead below and is not interruption safe with respect to consistency of the pool and `activeChannels` (which could be fixed, but another approach may be simpler).

Reply:
Done

Reply:
Yes this makes more sense to me too. We need to invalidate the item from the pool when the channel closed future completes. The promise is basically just a way to give us a self reference to the pool in the pool constructor.